src/plugins/fs/nilfs.c

```diff
@@ -0,0 +1,360 @@
+ *          successfully set or not
+ *
+ * Tech category: %BD_FS_TECH_NILFS2-%BD_FS_TECH_MODE_SET_UUID
+ */
+gboolean bd_fs_nilfs2_set_uuid (const gchar *device, const gchar *uuid, GError **error) {
+    const gchar *args[5] = {"nilfs-tune", "-U", uuid, device, NULL};
+    uuid_t uu;
+    gchar uuidbuf[37] = {0};
+
+    if (!uuid) {
+        uuid_generate (uu);
+        uuid_unparse (uu, uuidbuf);
+        args[2] = uuidbuf;
+    }
+
+    if (!check_deps (&avail_deps, DEPS_NILFSTUNE_MASK, deps, DEPS_LAST, &deps_check_lock, error))
+        return FALSE;
+
+    return bd_utils_exec_and_report_error (args, NULL, error);
+}
+
+/**
+ * bd_fs_nilfs2_check_uuid:
+ * @uuid: UUID to check
+ * @error: (out) (allow-none): place to store error
+ *
+ * Returns: whether @uuid is a valid UUID for the nilfs file system or not
+ *          (reason is provided in @error)
+ *
+ * Tech category: always available
+ */
+gboolean bd_fs_nilfs2_check_uuid (const gchar *uuid, GError **error) {
+    return check_uuid (uuid, error);
+}
+
+/**
+ * bd_fs_nilfs2_get_info:
+ * @device: the device containing the file system to get info for
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: (transfer full): information about the file system on @device or
+ *                           %NULL in case of error
+ *
+ * Tech category: %BD_FS_TECH_NILFS2-%BD_FS_TECH_MODE_QUERY
+ */
+BDFSNILFS2Info* bd_fs_nilfs2_get_info (const gchar *device, GError **error) {
+    const gchar *args[4] = {"nilfs-tune", "-l", device, NULL};
+    gboolean success = FALSE;
+    gchar *output = NULL;
+    BDFSNILFS2Info *ret = NULL;
+    gchar **lines = NULL;
+    gchar **line_p = NULL;
+    gchar *val_start = NULL;
+
+    if (!check_deps (&avail_deps, DEPS_NILFSTUNE_MASK, deps, DEPS_LAST, &deps_check_lock, error))
+        return FALSE;
+
+    ret = g_new0 (BDFSNILFS2Info, 1);
+
+    success = get_uuid_label (device, &(ret->uuid), &(ret->label), error);
+    if (!success) {
+        /* error is already populated */
+        bd_fs_nilfs2_info_free (ret);
+        return NULL;
+    }
+
+    success = bd_utils_exec_and_capture_output (args, NULL, &output, error);
+    if (!success)
+        /* error is already populated */
+        return FALSE;
+
+    lines = g_strsplit (output, "\n", 0);
+    g_free (output);
+    line_p = lines;
+
+    while (line_p && *line_p && !g_str_has_prefix (*line_p, "Block size:"))
+        line_p++;
+    if (!line_p || !(*line_p)) {
+        g_set_error (error, BD_FS_ERROR, BD_FS_ERROR_PARSE, "Failed to parse NILFS2 file system information");
+        g_strfreev (lines);
+        bd_fs_nilfs2_info_free (ret);
+        return FALSE;
+    }
+
+    /* * extract data from something like this: "Block size: 4096" */
+    val_start = strchr (*line_p, ':');
+    val_start++;
+    ret->block_size = g_ascii_strtoull (val_start, NULL, 0);
+
+    while (line_p && *line_p && !g_str_has_prefix (*line_p, "Device size"))
```

Comment:
Suggestion: what if the order of lines would change in the future? If you start from the beginning, there's better chance to find a match in such case.

Reply:
Ok, fixed. We don't do it for other filesystems, but we can start new tradition here :-)